### PR TITLE
Fix #1 BST Utility Working

### LINKED ISSUE #1: v1: Create BST Utility
# Version 1 Task Details
- [x] Build a Binary tree data structure with the following operations:
- [x] Add elements into the tree(multiple elements comma separated) 
- [x] Print the largest element 
- [x] Print the smallest element 
- [x] Print Inorder
- [x] Print Postorder
- [x] Print Level order
- [x] Print Preorder traversal
- [x] Search an element (by value)
- [x] Remove an element (by value)
- [x] Print all the paths i.e starting from root to the leaf 
- [x] Write all the elements to a file on quit 
- [x] Command Option to start the script with file input and load BST with integers inside that file

Note:
- Use OOPs concepts in this. 
- Follow coding guidelines 
- The script should run and serve until quit is not given as input

### PR DESCRIPTION
**Pull request by bv-ankit**
Adds all required functionalities for issue #1

### CODE REVIEW

File: bst.rb [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    ensure
+      output_file.close if output_file
+    end
+
+    puts "Exiting..."
+    break
+  when "1"
+    puts "Enter comma seperated numbers to add to the tree:"
+    new_input = STDIN.gets.chomp
+    new_input = new_input.split(",")
```
Comment (ritzzbv) [inline bst.rb:242]: best way to take integer input from STDIN

File: bst.rb [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+      output_file.puts(p bst.inorder_traversal)
+    rescue
+      puts "Error in writing file"
+    ensure
+      output_file.close if output_file
+    end
+
+    puts "Exiting..."
+    break
+  when "1"
```
Comment (ritzzbv) [inline bst.rb:239]: ADD_ELEMENT = add_operation(1, "MSG", "OUTPUT MSG", "ERROR MSG")
REMOVE_ELEMENT = add_operation(2, )

def add_operation
    OPERATION[1] = ["MSG", "OUTPUT MSG", "ERROR MSG"]
    1
end 

REMOVE_ELEMENT = 2

OPERATIONS = {}

ADD_ELEMENT => ["MSG", "OUTPUT MSG", "ERROR MSG"]

File: bst.rb [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    puts "Enter comma seperated numbers to add to the tree:"
+    new_input = STDIN.gets.chomp
+    new_input = new_input.split(",")
+    new_input.each { |str| abort "Invalid Input" unless str.to_i.to_s == str }
+    new_input = new_input.map { |curr_value| curr_value.to_i }
+    new_input.each {
+      |curr_value|
```
Comment (ritzzbv) [inline bst.rb:246]: new_input.each { |curr_value|

}

File: bst.rb (hunk 1 above)
Comment (bv-ankit) [reply bst.rb]: Checked, this is a common way.
[Link](https://stackoverflow.com/questions/31410694/how-to-check-if-user-input-is-a-number)

File: bst.rb (hunk 3 above)
Comment (bv-ankit) [reply bst.rb]: Done

File: lib/app.rb [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+def initialization()
+  puts "Welcome to Binary Search Tree Utility \n\n"
+  bst = BinarySearchTree.new()
+  puts "Do you want to initialize by importing values from a file ? \n" \
+       "Example: './output/output.txt' (Leave blank to not import)"
+  input_array = file_to_int_array
+  input_array.each { |value| bst.insert_node(value) }
+  puts "\n Initialization Complete! \n"
+  puts "Initial state of tree is: (Preorder)"
```
Comment (ritzzbv) [inline lib/app.rb:48]: count elements and show the count

File: lib/app.rb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+def file_to_int_array
+  file_location = get_file_location()
+  return [] if file_location == ""
+  input_string = read_from_file(file_location)
+  return [] if input_string == ""
+  input_array = convert_string_to_int_array(input_string)
+  input_array
+end
+
+def initialization()
```
Comment (ritzzbv) [inline lib/app.rb:40]:
```
module OperationCategory
     MODIFY = add_category(1, "msg")
end

module Operation
      LARGEST_ELEMENT = add_opertaion(1, "msg", OperationCategory::MODIFY)
end

operations = {}
add_category(type, msg)
   operations[type] = { :msg => msg}
     type
end

add_operation(type, msg, operation_type)
   operations[operation_type][type] = "msg"
   type
end
```

File: lib/app.rb [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+
+def perform_operation(curr_node, input_option)
+  option_selector = ''
+  curr_node["childs"].each { |option|
+    option_selector = option.perform if option.id == input_option
+    break
+  }
+
+  case option_selector
```
Comment (ritzzbv) [inline lib/app.rb:152]: case operation
when PRINT_LARGEST_ELEMENT

File: lib/classes/bst_node.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class Node
+  attr_accessor :value, :left_node, :right_node
+
+  def initialize(input_value, input_left_node: nil, input_right_node: nil)
```
Comment (ritzzbv) [inline lib/classes/bst_node.rb:4]: no need to prefix input

File: lib/classes/bst.rb [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  def inorder_rec(node)
+    return [] unless node
+    inorder_rec(node.left_node) +
+      [node.value] +
+      inorder_rec(node.right_node)
+  end
+
+  def preorder_rec(node)
+    return [] unless node
+    [node.value] +
```
Comment (ritzzbv) [inline lib/classes/bst.rb:161]: no need to make _rec functions. just use preorder_traversal

File: lib/app.rb [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+    break
+  when "INSERT_MULTIPLE"
+    new_input = take_multiple_inputs()
+    new_input.each { |curr_value|
+      if curr_value
+        bst.insert_node(curr_value)
+      end
+    }
+  when "LARGEST"
+    puts bst.find_largest_node().value
```
Comment (ritzzbv) [inline lib/app.rb:164]: don't use empty ()

File: lib/app.rb [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    quit_operation()
+    break
+  when "INSERT_MULTIPLE"
+    new_input = take_multiple_inputs()
+    new_input.each { |curr_value|
+      if curr_value
+        bst.insert_node(curr_value)
+      end
+    }
+  when "LARGEST"
```
Comment (ritzzbv) [inline lib/app.rb:163]: when SubOperation

File: lib/app.rb [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+  curr_node['childs'].each{ | options|
+    puts "#{options.id}. #{option['messages']['MSG']}"
+  }
+end
+
+def quit_operation()
+  puts "End state of BST(Inorder Traversal) saved in output.txt"
+  begin
+    output_file = File.new("output.txt", "w")
```
Comment (ritzzbv) [inline lib/app.rb:97]: read and write in binary mode

File: lib/app.rb [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+  file_location
+end
+
+def read_from_file(file_location)
+  begin
+    input_file = File.new(file_location, "r")
+    input_string = ""
+    input_file.each_byte { |ch| input_string << ch }
+  rescue => e
+    puts e.to_s
```
Comment (ritzzbv) [inline lib/app.rb:16]: puts "ERROR: READ_FROM_FILE: #{e}"

File: lib/app.rb [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+  input_array = file_to_int_array
+  input_array.each { |value| bst.insert_node(value) }
+  puts "\n Initialization Complete! \n"
+  puts "Initial state of tree is: (Preorder)"
+  p bst.preorder_traversal()
+  puts "\n\n"
+end
+
+
+def take_single_input()
```
Comment (ritzzbv) [inline lib/app.rb:54]: no need of empty ()

File: lib/app.rb [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    else
+      puts "Invalid Input!"
+      new_input = take_single_input()
+    end
+    new_input
+end
+
+def take_multiple_inputs()
+  puts "Enter comma seperated numbers:"
+  new_input = STDIN.gets.chomp
```
Comment (ritzzbv) [inline lib/app.rb:68]: just input

File: lib/app.rb [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+  new_input = new_input.split(",")
+  input_valid = true
```
Comment (ritzzbv) [inline lib/app.rb:70]: is_valid_input

File: lib/app.rb [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    option_selector = option.perform if option.id == input_option
+    break
+  }
+
+  case option_selector
+  when "QUIT"
+    quit_operation()
+    break
+  when "INSERT_MULTIPLE"
+    new_input = take_multiple_inputs()
```
Comment (ritzzbv) [inline lib/app.rb:157]: in case of multiple_input error, raise error from that function and handle this here.

File: lib/app.rb [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+    puts "Enter a value to remove:"
+    new_input = take_single_input()
+    puts " Removed #{bst.remove_by_value(new_input)} element."
+  when "ROOT_TO_LEAF"
+    bst.print_root_to_leaf_paths()
+  when "LEVELORDER"
+    p "Output: #{bst.levelorder_traversal()}"
+  else
+    puts "Invalid Input"
+  end
```
Comment (ritzzbv) [inline lib/app.rb:187]: check for invalid input up front
puts operation[type]["input_msg"]

Comment (ritzzbv) [reply lib/app.rb:187]: puts "#{operations[type]["outputmsg"]} : #{output}"

File: lib/app.rb [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+  new_input = STDIN.gets.chomp
+  new_input = new_input.split(",")
+  input_valid = true
+  new_input.each { |str| 
+    unless str.to_i.to_s == str 
+      puts "Invalid Input!"
+      input_valid = false
+      break
+    end
+  }
```
Comment (ritzzbv) [inline lib/app.rb:77]: input =is_valid_input ? input.map{|value| value.to_i} : take_multiple_inputs

File: lib/app.rb (hunk 4 above)
Comment (bv-ankit) [reply lib/app.rb:48]: Done

File: lib/classes/bst_node.rb (hunk 7 above)
Comment (bv-ankit) [reply lib/classes/bst_node.rb:4]: Done

File: lib/app.rb (hunk 9 above)
Comment (bv-ankit) [reply lib/app.rb:135]: Resolved

File: lib/app.rb (hunk 12 above)
Comment (bv-ankit) [reply lib/app.rb:16]: Resolved

File: lib/app.rb (hunk 13 above)
Comment (bv-ankit) [reply lib/app.rb:54]: Resolved

File: lib/app.rb (hunk 14 above)
Comment (bv-ankit) [reply lib/app.rb:68]: Resolved

File: lib/app.rb (hunk 15 above)
Comment (bv-ankit) [reply lib/app.rb:70]: Resolved

File: lib/app.rb (hunk 16 above)
Comment (bv-ankit) [reply lib/app.rb:128]: Resolved

File: lib/app.rb (hunk 18 above)
Comment (bv-ankit) [reply lib/app.rb:77]: Resolved

File: lib/app.rb (hunk 11 above)
Comment (bv-ankit) [reply lib/app.rb]: Writing char by char from string using each_char method to the file.
Writing using each_byte gives their ascii values instead.

Comment (bv-ankit) [reply lib/app.rb]: Oh, got it. Will use "rb", "wb" mode.

File: lib/classes/bst.rb (hunk 8 above)
Comment (bv-ankit) [reply lib/classes/bst.rb]: Resolved

File: lib/app.rb (hunk 6 above)
Comment (bv-ankit) [reply lib/app.rb]: Resolved

File: lib/app.rb (hunk 5 above)
Comment (bv-ankit) [reply lib/app.rb]: Resolved

File: bst.rb (hunk 2 above)
Comment (bv-ankit) [reply bst.rb]: Resolved

File: lib/app.rb [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+end
+
+$operations = {}
+
+def add_category(type, msg)
+  $operations[type] = { :msg => msg }
+  type
+end
+
+def add_operation(type, msg, operation_type)
```
Comment (ritzzbv) [inline lib/app.rb:190]: category not operation_type

File: lib/app.rb [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+
+  while true
+    selected_category ? display_operations(operations, selected_category) : display_categories(operations)
+
+    puts "Enter 'home' to go back to main menu."
+    puts "Enter 'quit' to exit."
+
+    input_option = STDIN.gets.chomp
+    puts "Input provided: #{input_option}"
+
```
Comment (ritzzbv) [inline lib/app.rb:235]:
```
while true
  if input == Commands::QUIT
    quit_operation(bst)
    break
  end

  if category
    if input == Commands::HOME
      category = nil
    else
      input = input.to_i
      puts "Invalid Operation" if !OPERATIONS.include?(input)

      perform_operation(input, category, bst, operations)
    end
  else
    category = input.to_i
  end
end
```

File: lib/app.rb [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+  QUIT = "quit"
+  HOME = "home"
+end
+
+module OperationCategory
+  PRINT = add_category(1, "Print")
+  MODIFY = add_category(2, "Modify")
+  SEARCH = add_category(3, "Search")
+end
+
```
Comment (ritzzbv) [inline lib/app.rb:205]: OPERATIONS = []

File: lib/app.rb [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+      new_input = take_single_input
+      puts "Search found at: #{bst.search_by_value(new_input)}"
+    end
+  else
+    puts "Invalid Input"
+  end
+  puts "\n\n"
+end
+
+$operations = {}
```
Comment (ritzzbv) [inline lib/app.rb:183]: OPERATIONS = {}

File: lib/app.rb (hunk 20 above)
Comment (ritzzbv) [reply lib/app.rb:235]: for operation you can use this o.values.collect { |op| op.keys }.flatten

File: lib/app.rb [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+      output_string = output_string.delete_suffix(",")
+      output_string.each_char { |char| output_file << char }
+    rescue => e
+      puts "ERROR: WRITE_TO_FILE: #{e.to_s}"
+    ensure
+      output_file.close if output_file
+    end
+  end
+  puts "Exiting..."
+end
```
Comment (ritzzbv) [inline lib/app.rb:116]: def handle_print_operation
end

**Comment by bv-ankit** [comment]
Resolved issues.

**Comment by bv-ankit** [comment]
Merging to master
#1 Resolved.

**Comment by bv-ankit** [comment]
Merged the branch.  @ritzzbv Thank you for guidance.

**Pull request merged**